src/path_tracking_pid_local_planner.cpp

```diff
@@ -0,0 +1,644 @@
+  for (const geometry_msgs::Point point : collision_polygon)
+  {
+    mkCollisionHull.points.push_back(point);
+  }
+
+  mkCollision.markers.push_back(mkCollisionFootprint);
+  mkCollision.markers.push_back(mkCollisionHull);
+  if (n_steps > 0)
+  {
+    mkCollision.markers.push_back(mkSteps);
+    mkCollision.markers.push_back(mkPosesOnPath);
+  }
+  collision_marker_pub_.publish(mkCollision);
+
+  return max_cost;
+}
+
+uint32_t TrackingPidLocalPlanner::computeVelocityCommands(const geometry_msgs::PoseStamped& pose,
+                                                          const geometry_msgs::TwistStamped& velocity,
+                                                          geometry_msgs::TwistStamped& cmd_vel, std::string& message)
+{
+  if (!initialized_)
+  {
+    ROS_ERROR("path_tracking_pid has not been initialized, please call initialize() before using this planner");
+    active_goal_ = false;
+    return mbf_msgs::ExePathResult::NOT_INITIALIZED;
+  }
+  // TODO(Cesar): Use provided pose and odom
+  if (!computeVelocityCommands(cmd_vel.twist))
+  {
+    active_goal_ = false;
+    return mbf_msgs::ExePathResult::FAILURE;
+  }
+  cmd_vel.header.stamp = ros::Time::now();
+  cmd_vel.header.frame_id = base_link_frame_;
+
+  bool moving = std::abs(cmd_vel.twist.linear.x) > VELOCITY_EPS;
+  if (cancel_in_progress_)
+  {
+    if (!moving)
+    {
+        ROS_INFO("Cancel requested and we now (almost) reached velocity 0: %f", cmd_vel.twist.linear.x);
+        cancel_in_progress_ = false;
+        active_goal_ = false;
+        return mbf_msgs::ExePathResult::CANCELED;
+    }
+    ROS_INFO_THROTTLE(1.0, "Cancel in progress... remaining x_vel: %f", cmd_vel.twist.linear.x);
+    return TrackingPidLocalPlanner::GRACEFULLY_CANCELLING;
+  }
+
+  if (!moving && pid_controller_.getVelMaxObstacle() < VELOCITY_EPS)
+  {
+    active_goal_ = false;
+    return mbf_msgs::ExePathResult::BLOCKED_PATH;
+  }
+
+  if (isGoalReached())
+    active_goal_ = false;
+  return mbf_msgs::ExePathResult::SUCCESS;
+}
+
+bool TrackingPidLocalPlanner::isGoalReached()
+{
+  // Return reached boolean, but never succeed when we're preempting
+  return pid_controller_.getControllerState().end_reached && !cancel_in_progress_;
+}
+
+bool TrackingPidLocalPlanner::isGoalReached(double dist_tolerance, double angle_tolerance)
+{
+  return isGoalReached();
+}
+
+bool TrackingPidLocalPlanner::cancel()
+{
+  // This function runs in a separate thread
+  cancel_requested_ = true;
+  cancel_in_progress_ = true;
+  ros::Rate r(10);
+  ROS_INFO("Cancel requested, waiting in loop for cancel to finish");
+  while (active_goal_)
+  {
+      r.sleep();
+  }
+  ROS_INFO("Finished waiting loop, done cancelling");
+  return true;
+}
+
+void TrackingPidLocalPlanner::curOdomCallback(const nav_msgs::Odometry& odom_msg)
+{
+  latest_odom_ = odom_msg;
+}
+
+void TrackingPidLocalPlanner::velMaxExternalCallback(const std_msgs::Float64& msg)
+{
+  pid_controller_.setVelMaxExternal(msg.data);
+}
```

Comment:
Yes, but this feature is for high-dynamic (thus streaming) maximum velocities.

If you want a guaranteed reception because of the (semi-)static nature of the limits, you can simply use the dynamic reconfigure (which is a service).

Reply:
Is this high-dynamic situation really needed? is dynamic reconfigure not dynamic-enough for the intended use?

Reply:
Unlike most robots, in Harvey we can run at full capacity of the drive-train. Since the power-input has a maximum, we have to limit the power-output. Slowing down is the easiest (and desired) way to accomplish this.

Due to the dynamic nature of our working-force, we adapt the output velocity at a 10hz rate.
Services are typically not meant for these kind or rates.